src/hotspot/share/opto/vtransform.cpp

```diff
@@ -902,44 +892,43 @@ VTransformApplyResult VTransformLoadVectorNode::apply(VTransformApplyState& appl
     }
   }
 
-  LoadVectorNode* vn = LoadVectorNode::make(opc, ctrl, mem, adr, adr_type, vlen, bt,
+  LoadVectorNode* vn = LoadVectorNode::make(sopc, ctrl, mem, adr, _adr_type, vlen, bt,
                                             control_dependency());
   DEBUG_ONLY( if (VerifyAlignVector) { vn->set_must_verify_alignment(); } )
   register_new_node_from_vectorization_and_replace_scalar_nodes(apply_state, vn);
   return VTransformApplyResult::make_vector(vn, vlen, vn->memory_size());
 }
 
 VTransformApplyResult VTransformStoreVectorNode::apply(VTransformApplyState& apply_state) const {
+  int     sopc = scalar_opcode();
+  uint    vlen = vector_length();
   StoreNode* first = nodes().at(0)->as_Store();
-  uint  vlen = nodes().length();
-  Node* ctrl = first->in(MemNode::Control);
+  Node* ctrl = apply_state.transformed_node(in_req(MemNode::Control));
+  // first has the correct memory state, determined by VTransformGraph::apply_memops_reordering_with_schedule
   Node* mem  = first->in(MemNode::Memory);
-  Node* adr  = first->in(MemNode::Address);
-  int   opc  = first->Opcode();
-  const TypePtr* adr_type = first->adr_type();
+  Node* adr  = apply_state.transformed_node(in_req(MemNode::Address));
 
   Node* value = apply_state.transformed_node(in_req(MemNode::ValueIn));
-  StoreVectorNode* vn = StoreVectorNode::make(opc, ctrl, mem, adr, adr_type, value, vlen);
+  StoreVectorNode* vn = StoreVectorNode::make(sopc, ctrl, mem, adr, _adr_type, value, vlen);
   DEBUG_ONLY( if (VerifyAlignVector) { vn->set_must_verify_alignment(); } )
   register_new_node_from_vectorization_and_replace_scalar_nodes(apply_state, vn);
   return VTransformApplyResult::make_vector(vn, vlen, vn->memory_size());
 }
 
 void VTransformVectorNode::register_new_node_from_vectorization_and_replace_scalar_nodes(VTransformApplyState& apply_state, Node* vn) const {
   PhaseIdealLoop* phase = apply_state.phase();
-  Node* first = nodes().at(0);
-
-  register_new_node_from_vectorization(apply_state, vn, first);
+  register_new_node_from_vectorization(apply_state, vn);
 
   for (int i = 0; i < _nodes.length(); i++) {
     Node* n = _nodes.at(i);
     phase->igvn().replace_node(n, vn);
   }
 }
 
-void VTransformNode::register_new_node_from_vectorization(VTransformApplyState& apply_state, Node* vn, Node* old_node) const {
+void VTransformNode::register_new_node_from_vectorization(VTransformApplyState& apply_state, Node* vn) const {
   PhaseIdealLoop* phase = apply_state.phase();
-  phase->register_new_node_with_ctrl_of(vn, old_node);
+  // The control is not always correct, but we set major_progress anyway.
+  phase->register_new_node(vn, apply_state.vloop().cl());
   phase->igvn()._worklist.push(vn);
   VectorNode::trace_new_vector(vn, "AutoVectorization");
```

Comment:
Removing the argument here allows us yet another removal of dependency on the old scalar graph. We only needed it for using the same control as the old graph - but that is not necessary, we can just use the CountedLoop as control, which is good enough.

Reply:
> we can just use the CountedLoop as control, which is good enough

For my understanding: this is because we can only vectorize if there are no other control dependencies in the loop?

Reply:
I think it should generally still be correct but might not need to be as accurate as possible. You'll never know if some code will rely on the correctness later in the same loop opts - might not today but at some point, especially when trying to add some verification code. So, IIUC, you are just more conservative/less accurate now while still being correct. Maybe you can tweak the comment to express that since "not always correct" could also imply some actual illegal control.

Reply:
Adjusted the comment.